core/scale/encoder/primitives.hpp

```diff
@@ -0,0 +1,434 @@
+      return;
+    }
+
+    if (val >= (1ul << 63ull)) {
+      encode(func, ::scale::CompactInteger{val});
+      return;
+    }
+
+    const size_t bigIntLength = sizeof(uint64_t) - (__builtin_clzll(val) / 8);
+
+    uint8_t result[sizeof(uint64_t) + sizeof(uint8_t)];
+    result[0] = (bigIntLength - 4) * 4 + 3;  // header
+
+    *(uint64_t *)&result[1] = math::toLE(val);
+    putByte(func, result, bigIntLength + 1ull);
+  }
+
+  template <typename F>
+  constexpr void encode(const F &func, const std::string &v) {
+    encode(func, std::string_view{v});
+  }
+
+  template <typename F>
+  constexpr void encode(const F &func, const std::string_view &v) {
+    encodeCompact(func, v.size());
+    putByte(func, (const uint8_t *)v.data(), v.size());
+  }
+
+  template <typename F>
+  void encode(const F &func, const ::scale::BitVec &v) {
+    const size_t bitsCount = v.bits.size();
+    const size_t bytesCount = ((bitsCount + 7ull) >> 3ull);
+    const size_t blocksCount = ((bytesCount + 7ull) >> 3ull);
+
+    encodeCompact(func, bitsCount);
+    uint64_t result;
+    size_t bitCounter = 0ull;
+    for (size_t ix = 0ull; ix < blocksCount; ++ix) {
+      result = 0ull;
+      size_t remains = std::min(size_t(64ull), bitsCount - bitCounter);
+      do {
+        result |= ((v.bits[bitCounter] ? 1ull : 0ull) << (bitCounter % 64ull));
+        ++bitCounter;
+      } while (--remains);
+
+      const size_t bits = (bitCounter % 64ull);
+      const size_t bytes =
+          (bits != 0ull) ? ((bits + 7ull) >> 3ull) : sizeof(result);
+
+      result = math::toLE(result);
+      putByte(func, (uint8_t *)&result, bytes);
+    }
+  }
+
+  template <typename F>
+  void encode(const F &func, const ::scale::CompactInteger &value) {
+    if (value < 0) {
+      raise(::scale::EncodeError::NEGATIVE_COMPACT_INTEGER);
+    }
+
+    if (value < ::scale::compact::EncodingCategoryLimits::kMinUint16) {
+      encodeCompactSmall(func, value.convert_to<uint8_t>());
+      return;
+    }
+
+    if (value < ::scale::compact::EncodingCategoryLimits::kMinUint32) {
+      encodeCompactSmall(func, value.convert_to<uint16_t>());
+      return;
+    }
+
+    if (value < ::scale::compact::EncodingCategoryLimits::kMinBigInteger) {
+      encodeCompactSmall(func, value.convert_to<uint32_t>());
+      return;
+    }
+
+    constexpr size_t kReserved = 68ull;
+    const size_t bigIntLength = countBytes(value);
+
+    if (bigIntLength >= kReserved) {
+      raise(::scale::EncodeError::COMPACT_INTEGER_TOO_BIG);
+    }
+
+    uint8_t result[kReserved];
+    result[0] = (bigIntLength - 4) * 4 + 3;  // header
+
+    namespace mp = boost::multiprecision;
+    const size_t size = value.backend().size();
+    const mp::limb_type *const p = value.backend().limbs();
+
+    constexpr size_t limb_sz = sizeof(mp::limb_type);
+    size_t ix = 0ull;
+    for (; ix < size; ++ix) {
+      *(mp::limb_type *)&result[1ull + ix * limb_sz] = math::toLE(p[ix]);
+    }
```

Comment:
May write comment that this is faster than
`export_bits(value, result + 1, 8, false);`